chain/cardano/tx/tx.go

```diff
@@ -0,0 +1,483 @@
+
+		for assetName, tokenAmount := range amounts {
+			contract := xc.ContractAddress(fmt.Sprintf("%s%s", hexPolId, assetName))
+			totalOutputs[contract] = xc.NewAmountBlockchainFromUint64(tokenAmount)
+		}
+	}
+
+	// Iterate over inputs and calculate change amounts
+	changeOutput, err := NewOutput(args.GetFrom())
+	if err != nil {
+		return nil, fmt.Errorf("failed to create change output: %w", err)
+	}
+	zeroAmount := xc.NewAmountBlockchainFromUint64(0)
+	for contract, inputAmount := range inputAmounts {
+		outputAmount, ok := totalOutputs[contract]
+		if !ok {
+			outputAmount = zeroAmount
+		}
+		log.WithFields(log.Fields{
+			"contract": contract,
+			"amount":   outputAmount,
+		}).Debug("output amount")
+
+		changeAmount := inputAmount.Sub(&outputAmount)
+		if changeAmount.Cmp(&zeroAmount) == -1 {
+			return nil, fmt.Errorf("negative change amount for contract %s", contract)
+		}
+		changeOutput.AddAmount(changeAmount.Uint64(), contract)
+		log.WithFields(log.Fields{
+			"contract": contract,
+			"amount":   changeAmount,
+		}).Debug("change amount")
+	}
+
+	return changeOutput, nil
+}
+
+// Create Cardano transaction
+// 1. Create raw transaction
+// 2. Create inputs
+// 3. Create recipient output
+// 4. Create change output
+// 5. Create dummy signature for fee estimation
+// 6. Deduct fee from change output
+func NewTx(args xcbuilder.TransferArgs, input tx_input.TxInput) (xc.Tx, error) {
+	tx := newTx()
+
+	inputMemo, ok := args.GetMemo()
+	if ok {
+		tx.Memo = &inputMemo
+		bytes, err := cbor.Marshal(inputMemo)
+		if err != nil {
+			return nil, errors.New("failed to marshal memo")
+		}
+		hash := blake2b.Sum256(bytes)
+		tx.Body.MemoHash = hash[:]
+	}
+
+	// Create inputs
+	for _, utxo := range input.Utxos {
+		hexHash, err := hex.DecodeString(utxo.TxHash)
+		if err != nil {
+			return nil, fmt.Errorf("failed to decode hash: %w", err)
+		}
+		txInput := &Input{
+			TxHash: hexHash,
+			Index:  utxo.Index,
+		}
+		tx.Body.Inputs = append(tx.Body.Inputs, txInput)
+	}
+
+	// Recipient output
+	output, err := CreateOutput(args, input)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create output: %w", err)
+	}
+	tx.Body.Outputs = append(tx.Body.Outputs, output)
+
+	// Change output
+	changeOutput, err := CreateChangeOutput(input.Utxos, output, args)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create change output: %w", err)
+	}
+	tx.Body.Outputs = append(tx.Body.Outputs, changeOutput)
+
+	tx.Body.TTL = uint32(input.Slot + 3600*2) // 2 hours
+
+	// Witness for fee estimation
+	dummyWitness := &VKeyWitness{
+		VKey:      make([]byte, 32),
+		Signature: make([]byte, 64),
+	}
+	tx.Witness.Keys = append(tx.Witness.Keys, dummyWitness)
+	// Make sure to zero out the dummy witness
+	defer func() {
+		tx.Witness.Keys = make([]*VKeyWitness, 0)
+	}()
+
+	// Serialize tx with dummy signature for fee estimation
+	txCbor, err := cbor.Marshal(tx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal tx: %w", err)
+	}
```

Comment:
This could be done in the client itself in FetchTransferInput, since it would have access to both `args xcbuilder.TransferArgs, input tx_input.TxInput`.  I recommend you refactor so that the Client makes it's own dummy cardano Tx for fee estimation (perhaps using 0 fee initially).

Then the real transaction builder can use the size-based estimation.